### PR TITLE
NTLM MIC calculation draft

### PR DESCRIPTION
**Pull request by sopak**
We run into an issue where some server type (EMC Unity) fail when MIC was presented and not requested.
Another issue with another type (NetApp 7mode) if MIC was presented and not calculated correctly.

MIC (16 bytes) provided and offset is modified if in AvPairType is key MsvAvFlags with value & 0x00000002 is true
otherwise, MIC is skipped and offset is not modified.
MIC calculation is a draft and needs further testing.
Documentation on address https://msdn.microsoft.com/en-us/library/cc236695.aspx
is not very clear.

### CODE REVIEW

File: src/main/java/com/hierynomus/ntlm/messages/NtlmAuthenticate.java [hunk 1]
```diff
@@ -15,32 +15,34 @@
...
     private static byte[] EMPTY = new byte[0];
-
     private byte[] lmResponse;
     private byte[] ntResponse;
     private byte[] userName;
     private byte[] domainName;
     private byte[] workstation;
     private byte[] encryptedRandomSessionKey;
     private long negotiateFlags;
+    private boolean useMIC;
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/ntlm/messages/NtlmAuthenticate.java]: Though MIC is an acronym, please rename to `useMic`, and also the field below to `mic` (as `MIC` is a constant name)

File: src/main/java/com/hierynomus/smbj/auth/NtlmAuthenticator.java [hunk 2]
```diff
@@ -87,27 +83,52 @@ public NtlmAuthenticator create() {
...
-                    sessionkey = ntlmFunctions.encryptRc4(userSessionKey, masterKey);
-                    session.setSigningKey(masterKey);
-                } else {
-                    sessionkey = userSessionKey;
-                    session.setSigningKey(sessionkey);
-                }
+            byte[] sessionkey;
+
+            byte[] userSessionKey = ntlmFunctions.hmac_md5(responseKeyNT, Arrays.copyOfRange(ntlmv2Response, 0, 16)); // first 16 bytes of ntlmv2Response is ntProofStr
+            if (challenge.getNegotiateFlags().contains(NtlmNegotiateFlag.NTLMSSP_NEGOTIATE_KEY_EXCH)
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/auth/NtlmAuthenticator.java]: extract `challenge.getNegotiateFlags` to a local variable to make the if-statement more readable.

File: src/main/java/com/hierynomus/smbj/auth/NtlmAuthenticator.java [hunk 3]
```diff
@@ -87,27 +83,52 @@ public NtlmAuthenticator create() {
...
+                );
+
+                // TODO correct hash should be tested
+
+                Buffer.PlainBuffer concatenatedBuffer = new Buffer.PlainBuffer(Endian.LE);
+                concatenatedBuffer.putRawBytes(negTokenTarg.getResponseToken()); //negotiateMessage
+                concatenatedBuffer.putRawBytes(challenge.getServerChallenge()); //challengeMessage
+                resp.writeAutentificateMessage(concatenatedBuffer); //authentificateMessage
+
+                byte[] MIC = ntlmFunctions.hmac_md5(userSessionKey, concatenatedBuffer.getCompactData());
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/auth/NtlmAuthenticator.java]: same here, rename to lowercase `mic`

**Comment by hierynomus** [comment]
@sopak Let me know when you think the PR is ready. For now I'll attach a label to it to mark it in progress.

**Comment by sopak** [comment]
We can merge it.
It will not cover all possibilities and exotic settings,
but I tested it with the following list of servers, and  I can authenticate:

Windows 10
Isilon 8
Netapp 7mode
Netapp CDOT
VNX 8
Unity
Qnap

Someone else can fine grained it in the future.

**Pull request merged**